MudList: Generic MudList<T> and other improvements

henon opened the pull request:
## Description
Resolves #1295 

Changes:
- `MudList` and `MudListItem` are now generic `<T>`. The only downside is that you must specify `T` now, even if you don't use the selection feature.
- `SelectedItem` is no longer a parameter and has been made internal. Selection is solely managed via `SelectedValue`
- **MudListItem**: Removed `InitiallyExpanded`, just set `Expanded` to `true`
- **MudListItem**: For `T="string"`, you can just set `Text` if the `Value` is the same so you don't have to repeat yourself. This also means you don't have to set a value if you only have a (unique) `Text` and you won't get selection problems. In the old list you had to set a unique `Value` just so the selection mechanism worked.
- Renamed `Clickable` to `ReadOnly` and inverted conditions. 
- Removed any pre-conditions for selection to work, such as binding `SelectedValue`. Lists are interactive by default now. If the list is `ReadOnly`, the selection will still be shown but can't be changed, if `Disabled` no selection shows.
- Renamed `DisablePadding` to `Padding` and inverted conditions
- Docs have been improved

![image](https://github.com/MudBlazor/MudBlazor/assets/44090/dd1a65f8-5177-43c8-a387-c243af924a0c)
![image](https://github.com/MudBlazor/MudBlazor/assets/44090/13a20f14-3cf5-4841-aacd-80a4c881c483)
![image](https://github.com/MudBlazor/MudBlazor/assets/44090/9a9fd7a3-6f9b-4a58-af01-cee9bb4169df)
![image](https://github.com/MudBlazor/MudBlazor/assets/44090/5520e438-7f31-4d17-aed0-e18c9b27b1ee)

Not in this PR but planned for another PR:
- Remove `string Avatar` in favor of `AvatarContent` render fragment
- Add `ListHeader` and `ListFooter` render fragments that don't scroll. Current solution `<ListSubHeader>` is nothing but a non-selectable item that scrolls out of view. 
- Maybe add support for second and third line like Vuetify has in MudListItem (if @danielchalmers helps with the CSS)

## How Has This Been Tested?
<!-- All PRs should implement unit tests if possible -->
<!-- Please describe how you tested your changes. -->
<!-- Have you created new tests or updated existing ones? -->
<!-- e.g. unit | visually | none -->

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

<!-- If you made any visual changes, provide screenshots of before/after. If it has moving parts, please provide high quality gif, webm, or mp4 -->

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [ ] I've added relevant tests.

henon committed:
wip

henon committed:
Merge remote-tracking branch '_MudBlazor/dev' into generic-list

# Conflicts:
#	src/MudBlazor.Docs/Pages/Components/ColorPicker/Examples/ColorPickerExampleUsageExample.razor
#	src/MudBlazor.Docs/Pages/Components/List/Examples/ListSelectionExample.razor
#	src/MudBlazor/Components/List/MudList.razor.cs
#	src/MudBlazor/Components/List/MudListItem.razor
#	src/MudBlazor/Components/List/MudListItem.razor.cs

henon committed:
fixes

henon committed:
refactoring

henon committed:
Improve Docs

henon committed:
Improve Docs

henon committed:
Merge remote-tracking branch '_MudBlazor/dev' into generic-list

henon committed:
update

henon committed:
argh

ScarletKuro commented:
You can convert `OnClickHandlerPreventDefault` to autoproperty:
```diff
-        private bool _onClickHandlerPreventDefault = false;


        [Parameter]
        [Category(CategoryTypes.List.Behavior)]
-        public bool OnClickHandlerPreventDefault
-        {
-            get => _onClickHandlerPreventDefault;
-            set => _onClickHandlerPreventDefault = value;
-        }
+        public bool OnClickHandlerPreventDefault { get; set; } = false;

...

-            if (!_onClickHandlerPreventDefault)
+            if (!OnClickHandlerPreventDefault)
            {
                if (NestedList != null)
                {
```

henon commented:
I am sure I can do something about `_textTypo` as well

henon committed:
minor fix

Anu6is commented on src/MudBlazor.Docs.Server/Properties/launchSettings.json:
```diff
@@ -13,7 +13,7 @@
       "launchBrowser": true,
       "applicationUrl": "https://localhost:7128;http://localhost:5128",
-      "launchUrl": "components/togglegroup",
+      "launchUrl": "components/list",
```
launchUrl can probably be left blank

henon replied:
Actually, I'd rather leave it. It helps a great deal while testing. Because you'll restart the docs 69 times and it saves you a few clicks every time. If I remove it, next time I need to look it up again and remove again.

ScarletKuro replied:
I think we should just add this file to .gitignore, because I see every new component this file is getting modified :)
But I think this file should have a default configuration, for newcomers that fetch the docs.

Anu6is commented on src/MudBlazor.Docs/Components/LandingPage/MiniApp/MiniApp.razor:
```diff
@@ -304,8 +304,8 @@
     private bool expansionPanel1 = true;
-    MudListItem selectedItem;
-    object selectedValue = 1;
+    //MudListItem selectedItem;
```
commented out code can be deleted

henon committed:
one should never push too early

ScarletKuro commented:
We probably should add the `CascadingTypeParameter` too
https://learn.microsoft.com/en-us/aspnet/core/blazor/components/generic-type-support?view=aspnetcore-8.0#cascaded-generic-type-support

ScarletKuro commented:
Tho, there is one disadvantage:
```
Matching is only performed by name. Therefore, we recommend avoiding a cascaded generic type parameter with a generic name, for example T or TItem. If a developer opts into cascading a type parameter, they're implicitly promising that its name is unique enough not to clash with other cascaded type parameters from unrelated components.
```
which I was talking before, I was told that changing the generic name of `T` to something else will break the docs

ScarletKuro commented on src/MudBlazor/Components/List/MudListItem.razor.cs:
```diff
@@ -245,7 +228,7 @@ protected async Task OnClickHandlerAsync(MouseEventArgs eventArgs)
         {
             await base.OnInitializedAsync();
-            await _expandedState.SetValueAsync(InitiallyExpanded);
+            //await _expandedState.SetValueAsync(InitiallyExpanded);
```
Remove the commented code

ScarletKuro reviewed (approved):
LGTM. I don't have anything else to offer, it's good enough for v7, the rest can be moved to v8.

Anu6is commented:
> We probably should add the `CascadingTypeParameter` too https://learn.microsoft.com/en-us/aspnet/core/blazor/components/generic-type-support?view=aspnetcore-8.0#cascaded-generic-type-support

This would be nice instead of having to repeat T="xxx" for each list item 😩

henon commented:
> This would be nice instead of having to repeat T="xxx" for each list item 😩

Yes. And thanks for reviewing!

henon commented:
> I was told that changing the generic name of `T` to something else will break the docs

Yeah, but we'll find a solution for that as well

henon committed:
update

mckaragoz commented:
1. I strongly offer to change Clickable and `@bind-SelectedValue` mechanism. For interactivity users have to set `@bind-SelectedValue` which can easily missed.
2. We should change `SelectedValue` to `Value`
3. Do we remove `SelectedItem` thing in v7? I think we should.

henon commented:
> * I strongly offer to change Clickable and `@bind-SelectedValue` mechanism. For interactivity users have to set `@bind-SelectedValue` which can easily missed. We should remove CanSelect thing https://github.com/henon/MudBlazor/blob/generic-list/src/MudBlazor/Components/List/MudList.razor.cs#L119

I agree. What if we make a parameter `ReadOnly` instead. So list will be interactive unless `ReadOnly="true"`

henon commented:
> 3\. Do we remove `SelectedItem` thing in v7? I think we should.

It is already removed

henon commented:
> 2\. We should change `SelectedValue` to `Value`

Not sure about that. We have `SelectedValue` in many other components that provide a selection, like for instance ChipSet, ButtonToggleGroup, Select, etc.

ScarletKuro commented:
> Not sure about that. We have `SelectedValue` in many other components that provide a selection, like for instance ChipSet, ButtonToggleGroup, Select, etc.

I guess this is open topic, since we had this PR https://github.com/MudBlazor/MudBlazor/pull/7892
that would remove the `Selected` prefix from radio etc.

mckaragoz commented:
> > Not sure about that. We have `SelectedValue` in many other components that provide a selection, like for instance ChipSet, ButtonToggleGroup, Select, etc.
> 
> I guess this is open topic, since we had this PR #7892 that would remove the `Selected` prefix from radio etc.

Yes i also remember to change all to `Value` (option, selectedvalue, etc.)

henon commented:
Ok, so renaming SelectedValue to Value is a thing for another PR and will have to be decided. I understand why Radio Checkbox etc were changed. But I am still not sure about ChipSet and List

mckaragoz commented:
And one advanced thing. Selected item rendering does not work properly when we use virtualize. It's very important feature. We should bring a new mechanism that check the item on each render and apply the selected style again without affect performance.

henon commented:
If that is not  a breaking change we can do it any time

mckaragoz commented:
Does list support SecondaryText and MultiSelection? These are key features in extended list.

henon commented:
I would love to add both. If you want we can do it in separate PRs. One for each feature please.

henon committed:
Remove all T="string" from MudListItems all over the code base due to cascading type param

henon committed:
Rename Clickable to ReadOnly and DisablePadding to Padding

henon commented:
@mckaragoz I renamed `Clickable` to `ReadOnly` and removed any conditions for selection to work. Only if read only selection will still be showed but can't be changed, if disabled no selection shows.

Also renamed `DisablePadding` to `Padding`

henon commented:
Hehe, and as is tradition for MudBlazor, MudList now also has a nuclear energy themed example

![image](https://github.com/MudBlazor/MudBlazor/assets/44090/abc4c27d-3a52-400e-b1d0-8eec7f89616b)

henon committed:
improve nuclear example

henon committed:
fix regression

henon committed:
ReadOnly list still shows the selected item as hilighted

henon committed:
Merge remote-tracking branch '_MudBlazor/dev' into generic-list

# Conflicts:
#	src/MudBlazor.Docs/Pages/Components/DropZone/Examples/DropZoneMailExample.razor
#	src/MudBlazor.Docs/Pages/Components/List/Examples/ListInteractiveExample.razor
#	src/MudBlazor/Components/List/MudListItem.razor.cs

henon committed:
minor change

henon committed:
Merge branch 'dev' into generic-list

henon committed:
Revert "Auxiliary commit to revert individual files from 7fc42108dc5fdfb2528001016780592a8ed3e325"

This reverts commit ce1950cfb907ddb56ed492edfb7ac6941719db61.

henon committed:
Merge remote-tracking branch 'henon/generic-list' into generic-list

henon commented:
Added to v7.0.0 Migration Guide #8447

haas-daniel commented on src/MudBlazor/Components/List/MudList.razor.cs:
```diff
@@ -43,21 +52,21 @@ public partial class MudList : MudComponentBase, IDisposable
         [Parameter]
         [Category(CategoryTypes.List.Appearance)]
-        public bool DisablePadding { get; set; }
+        public bool Padding { get; set; }
```
@henon did you forget to set the default value for Padding to true, or was that intentional?

henon replied:
It is an oversight, sorry. Interesting, that this hasn't been noticed sooner.

henon replied:
Maybe we should just keep it that way. To be honest, I never liked this extra bit of space at the bottom of the list which is a result of `Padding="true"`:

![image](https://github.com/user-attachments/assets/c526f9ad-6c16-4253-aedf-036849e8def4)

henon replied:
But of course, the header looks better with padding. Without:

![image](https://github.com/user-attachments/assets/a50179b7-b895-44dd-8491-0625f93bea13)

haas-daniel replied:
Yes, that would be ok for me. I don't need the extra space by default. I just saw it during testing.

henon replied:
Thanks, I'm adding this information to the migration guide #8447

ScarletKuro commented on src/MudBlazor/Components/List/MudListItem.razor.cs:
```diff
@@ -270,12 +253,7 @@ private void OnListParametersChanged()
-            }
-
-            if (_selected == selected)
+            if (GetDisabled() || _selected == selected)
```
I think this is causing a problem with preselected items when you have `Disabled=true`.
https://try.mudblazor.com/snippet/ckcyvFcEUQAxBtmA
I can create an issue for that.

ScarletKuro replied:
Nvm, issue was created already https://github.com/MudBlazor/MudBlazor/issues/10291
